src/basic/virt.c

```diff
@@ -181,8 +179,43 @@ static int detect_vm_dmi(void) {
                                 return dmi_vendor_table[j].id;
                         }
         }
-#endif
+        return VIRTUALIZATION_NONE;
+}
 
+static int detect_vm_dmi(void) {
+#if defined(__i386__) || defined(__x86_64__) || defined(__arm__) || defined(__aarch64__)
+
+        int r;
+        r = detect_vm_dmi_vendor();
+        if (r == VIRTUALIZATION_AMAZON) {
+                /* The DMI vendor tables in /sys/class/dmi/id don't help us distinguish between Amazon EC2
+                 * virtual machines and bare-metal instances, so we need to look deeper.  The EC2 firmware
+                 * BIOS Charateristics Extension Byte 2 (Section 2.1.2.2 of
+                 * https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.4.0.pdf), which
+                 * specifies that the 4th bit being set indicates a VM. The BIOS Characteristics table is
+                 * exposed via the kernel in /sys/firmware/dmi/entries/0-0. */
+                _cleanup_free_ char *s = NULL;
+                int i;
+                size_t readsize;
+                i = read_full_virtual_file("/sys/firmware/dmi/entries/0-0/raw", &s, &readsize);
+                if (i < 0) {
+                        log_debug("Unable to read /sys/firmware/dmi/entries/0-0/raw: %d", i);
+                        return r;
+                }
+                if (readsize < 20) {
+                        log_debug("Only read %lu bytes from /sys/firmware/dmi/entries/0-0/raw", readsize);
+                        return r;
+                }
+                unsigned byte = s[19];
+                if (byte ^ 1<<4) {
+                        log_debug("DMI BIOS Extension table does not indicate virtualization");
+                        return VIRTUALIZATION_NONE;
+                }
+                log_debug("DMI BIOS Extension table indicates virtualization");
+        }
```

Comment:
I'd prefer if the check for this bit could be split out into a function of its own, it's entirely generic and not specific to amazon after all. Morevoer, I think we should always call this at the end of our tests as a last-ditch effort to detect VM environments we otherwise don't recognize. And then return `VIRTUALIZATION_VM_OTHER`, to indicate that yes, we know we are in a VM, but no we have no clue in what.

Reply:
Are you sure you want to trust everybody's firmware to set that bit correctly?

Reply:
The latest revision moves this into a separate function. It's called in two places:
1. After the `/sys/class/dmi/id` checks identify an Amazon EC2 system, to validate that it is a VM rather than a bare-metal instance.
2. If no virtualization is detected, with return `VIRTUALIZATION_VM_OTHER` if this bit is set. I still have some concern that this might lead to hardware systems with buggy firmware reporting as VMs, but I don't know of any specific instances of such hardware.

Reply:
> Are you sure you want to trust everybody's firmware to set that bit correctly?

Well, given that the flag is kinda hidden in some auxiliary struct that mainboard vendors hopefully never bother with I'd at least try it. If it doesnt work, and we get reports of false positives, we can drop it again. But the only way to figure it out is to try.